architecture/2025-03-upgrade-design-doc.md

```diff
@@ -0,0 +1,677 @@
+    UCP -.->|"Only holds metadata"| UserApps["User Applications"]
+```
+
+### Detailed Design
+
+```mermaid
+graph TD
+    Start[User runs rad upgrade kubernetes] --> ParseArgs[Parse command arguments]
+    ParseArgs --> ValidateVersion[Validate version compatibility]
+    ValidateVersion --> AcquireLock[Acquire upgrade lock]
+    AcquireLock --> RunPreflights[Run pre-flight checks]
+    RunPreflights --> UserDataBackup[Create user data backup]
+    UserDataBackup --> PlanUpgrade[Calculate upgrade plan]
+    PlanUpgrade --> ExecuteHelmUpgrade[Execute Helm chart upgrade]
+    ExecuteHelmUpgrade --> MonitorProgress[Monitor upgrade progress]
+    MonitorProgress --> VerifyComponents[Verify component health]
+    VerifyComponents --> Success{Successful?}
+    Success -- Yes --> ReleaseLock[Release upgrade lock]
+    Success -- No --> UserDataRestore[Restore user data]
+    UserDataRestore --> RollbackHelm[Rollback Helm changes]
+    RollbackHelm --> ReleaseLock
+    ReleaseLock --> End[Display results to user]
+```
+
+**Version Detection and Validation:**
+
+The upgrade process begins with detecting the currently installed Radius version (the versions of the components in the cluster) and validating it against the requested target version. We can build an interface (or use/improve an existing one) to achieve this:
+
+```go
+// VersionValidator interface
+type VersionValidator interface {
+    ValidateTargetVersion(currentVersion, targetVersion string) error
+    GetLatestVersion() (string, error)
+    IsValidVersion(version string) bool
+}
+```
+
+This interface will be implemented (or existing will be improved) to handle version comparisons, prevent downgrades, and resolve the "latest" version tag to a specific version number. The implementation will (probably) connect to the GitHub API to fetch available release versions when needed.
+
+**Upgrade Lock Mechanism:**
+
+To prevent concurrent modifications during upgrades, we'll implement a lock system that's acquired early in the process:
+
+```go
+type UpgradeLock interface {
+    // Acquires the upgrade lock, preventing other commands from modifying data
+    AcquireLock(ctx context.Context) error
+
+    // Releases the upgrade lock, allowing other commands to proceed
+    ReleaseLock(ctx context.Context) error
+
+    // Checks if an upgrade is currently in progress
+    IsUpgradeInProgress(ctx context.Context) (bool, error)
+}
+```
+
+We can utilize Kubernetes Lease objects (coordination.k8s.io/v1) for implementing the distributed locking mechanism (open to discussion and suggestions). Leases are purpose-built for this use case, providing built-in lease duration and automatic expiration capabilities. For more information, see: <https://kubernetes.io/docs/concepts/architecture/leases/>.
+
+Other CLI commands (`rad deploy app.bicep`, `rad delete app my-app` or other data-changing commands) that modify data will check for this lock before proceeding:
```

Comment:
Locks should be enforced on the server side to prevent race conditions, as client-side locking may not effectively handle concurrent requests where client mutations and upgrades overlap.

Reply:
It is going to be moved to the data store layer.